onlinejudge/service/toph.py

```diff
@@ -0,0 +1,168 @@
+            log.debug('case: %s', str(case))
+            assert len(list(case.children))
+            input_pre, output_pre = list(map(lambda td: td.find('pre'), list(case.children)))
+            assert input_pre.name == 'pre'
+            assert output_pre.name == 'pre'
+            assert re.search("^preSample.*Input$", input_pre.attrs['id'])
+            assert re.search("^preSample.*Output$", output_pre.attrs['id'])
+            s = input_pre.get_text()
+            s = s.lstrip()
+            samples.add(s, "Input")
+            s = output_pre.get_text()
+            s = s.lstrip()
+            samples.add(s, "Output")
+        return samples.get()
+
+    def get_language_dict(self, session: Optional['requests.Session'] = None) -> Dict[str, onlinejudge.type.Language]:
+        session = session or utils.new_default_session()
+        # get
+        resp = utils.request('GET', self.get_url(), session=session)
+        # parse
+        soup = bs4.BeautifulSoup(resp.content.decode(resp.encoding), utils.html_parser)
+        select = soup.find('select', attrs={'name': 'languageId'})
+        if select is None:
+            log.error('not logged in')
+            return {}
+        language_dict = {}
+        for option in select.findAll('option'):
+            language_dict[option.attrs['value']] = {'description': option.string.strip()}
+        return language_dict
+
+    def submit_code(self, code: bytes, language: str, session: Optional['requests.Session'] = None) -> onlinejudge.type.Submission:  # or SubmissionError
+        session = session or utils.new_default_session()
+        # get
+        resp = utils.request('GET', self.get_url(), session=session)
+        # parse
+        soup = bs4.BeautifulSoup(resp.content.decode(resp.encoding), utils.html_parser)
+        form = soup.find('form')
+        if form is None:
+            log.error('not logged in')
+            raise SubmissionError
+        log.debug('form: %s', str(form))
+        if form.find('select') and form.find('select').attrs['name'] != 'languageId':
+            log.error("Wrong submission URL")
+            raise SubmissionError
+
+        # make data
+        form = utils.FormSender(form, url=resp.url)
+        form.set('languageId', language)
+        form.set_file('source', 'code', code)
+        resp = form.request(session=session)
+        resp.raise_for_status()
+        # result
+        if '/s/' in resp.url:
+            # example: https://toph.co/s/201410
+            log.success('success: result: %s', resp.url)
+            return onlinejudge.type.DummySubmission(resp.url)
+        else:
+            log.failure('failure')
+            log.debug('redirected to %s', resp.url)
+            raise SubmissionError
+
+    def get_url(self) -> str:
+        # TODO: Check for contest_id to return the appropriate  URL when support for contest is added
+        return f'https://toph.co/p/{self.problem_id}'
```

Comment:
```suggestion
        return 'https://toph.co/p/{}'.format(self.problem_id)
```

f-string is useful and you should use it if possible, however, we need to support Python 3.5, and unfortunately Python 3.5 doesn't support f-strings.